### DISCUSSION TITLE
Guardrails are now Free on LiteLLM Proxy

### DISCUSSION THREAD

**Discussion by ishaan-jaff** [BerriAI/litellm#5321] [Announcements]
# Guardrails are now Free on LiteLLM Proxy 

Guardrails are now free to use on LiteLLM Proxy -  use Aporia AI, Lakera AI 

Quick Start doc: https://docs.litellm.ai/docs/proxy/guardrails/quick_start

### Supported Integrations:
- Aporia AI https://docs.litellm.ai/docs/proxy/guardrails/aporia_api
- Lakera AI https://docs.litellm.ai/docs/proxy/guardrails/lakera_ai

### Supported Event hooks for guardrails
- `pre_call` Run **before** LLM call, on **input**
- `post_call` Run **after** LLM call, on **input & output**
- `during_call` Run **during** LLM call, on **input** Same as `pre_call` but runs in parallel as LLM call.  Response not returned until guardrail check completes

### What features will be in Enterprise for Guardrails?
- Controlling guardrails per API Key and Team: https://docs.litellm.ai/docs/proxy/guardrails/quick_start#-control-guardrails-per-project-api-key

![Group 5971](https://github.com/user-attachments/assets/a9f6ada0-65a7-42df-b885-5307afde5456)

**Comment by ryewilson** [comment]
@ishaan-jaff - question, is the `/guardrails/list` endpoint free to use as well? The comment in the docs indicates it's an enterprise feature (here: https://docs.litellm.ai/docs/proxy/guardrails/quick_start#1--view-available-guardrails)

**Comment by darfaz** [comment]
Great to see guardrails becoming a first-class feature in LiteLLM.

For anyone implementing guardrails in production, a few practical tips:

1. **Layer your guardrails** — LiteLLM's built-in guardrails are great for provider-specific checks (like Bedrock Guardrails), but you often need application-level rules too: PII detection, topic restrictions, output format validation. Running multiple guardrail layers (provider + application) gives defense in depth.

2. **Don't forget output guardrails** — Most teams focus on input scanning but model responses can also contain sensitive data, especially with RAG. Scanning outputs is equally important.

3. **Custom guardrail hooks** — For teams that need configurable input/output validation beyond what's built in, [ClawMoat](https://github.com/darfaz/clawmoat) provides a standalone guardrail layer that can work alongside LiteLLM. It handles things like prompt injection detection, PII scanning, and content policy enforcement with a simple API.

4. **Performance consideration** — Guardrails add latency. For real-time chat, consider running lightweight pattern-based checks synchronously and heavier ML-based checks asynchronously (log + alert rather than block).

Would love to see LiteLLM expose guardrail metrics (blocks, warnings, latency) through the existing observability integrations.
